README
Statocles::AppRole::ExtraFeeds

This module is a role that can be applied to any "Statocles::App" in a
"Statocles"'s "Beam::Wire" configuration.

  ...
  blog_app:
    class: 'Statocles::App::Blog'
    with: 'Statocles::AppRole::ExtraFeeds'
    args:
      url_root: '/blog'
      # ... more Statocles::App::Blog args
      extra_feeds:
        fulltext.rss:
          text: "RSS FullText"

This example creates a feed called "/blog/fulltext.rss" containing the
contents of "theme/blog/fulltext.rss.ep" after template application, and
is linked from every "index" listing.

It also creates a feed called "/blog/tag/<%= tagname %>.fulltext.rss"
for each tag, provisioned from the same template.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Statocles-AppRole-ExtraFeeds

0.001004 2017-03-07T09:19:21Z b81292b

0.001003 2017-03-07T09:17:28Z a29ca76
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot
 - Bugfix: Port tests to Statocles 0.75 changes

 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - test: ↑1 ↓1

0.001002 2016-02-26T06:52:21Z 8cbb246
 - Remove unused import in tests that will break in a future Statocles release

0.001001 2016-02-25T00:34:57Z 57745c3
 - Bump version on Beam::Wire for interop with Beam::Wire 'with' mechanics.

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - configure: +2
 - develop: +1
 - test: ↑1

0.001000 2016-02-24T15:13:45Z 55ebf9f
 - First version.


t/lib/KENTNL/Utils.pm
use 5.006;    # our
use strict;
use warnings;

package KENTNL::Utils;

our $VERSION = '0.001000';

# ABSTRACT: Output function lists heirachially

# AUTHORITY

use Exporter        ();
use Term::ANSIColor ('colored');
use Carp            ('croak');
use Package::Stash  ();

BEGIN { *import = \&Exporter::import }

our @EXPORT = qw( symdump has_feeds );

use constant {
  map { ( sprintf '_E%x', $_ ), ( sprintf ' E<%s#%d>', __PACKAGE__, $_ ) } 1 .. 4
};

{
  no strict;    # namespace clean
  delete ${ __PACKAGE__ . '::' }{ sprintf '_E%x', $_ } for 1 .. 4;
}

our @TYPE_METHOD      = ('cyan');
our @TYPE             = ('yellow');
our @PRIVATE          = ('reset');
our @PUBLIC           = ( 'bold', 'bright_green' );
our @SHADOWED_PRIVATE = ('magenta');
our @SHADOWED_PUBLIC  = ('red');

our $MAX_WIDTH     = 80;
our $SHOW_SHADOWED = 1;
our $INDENT        = q[ ] x 4;
our $CLUSTERING    = 'type_shadow_clustered';

sub symdump {
  my $nargs = scalar( my ( $target, ) = @_ );
  $nargs == 1     or croak "Passed $nargs arguments, Expected 1" . _E1;
  defined $target or croak "Expected defined target" . _E2;
  length $target  or croak "Expected target with non-zero length" . _E3;
  !ref $target    or croak "Expected scalar target" . _E4;
  _pp_key() . _pp_class($target);
}

sub has_feeds {
  my ( $desc, $page, @feeds ) = @_;
  require Test::More;
  my %has_feeds;
  for my $link (  $page->links('feed') ) {
    $has_feeds{ $link->href }++;
  }
  for my $feed (@feeds) {
    $has_feeds{$feed}--;
  }
  Test::More::note("Feeds for $desc: ");
  return Test::More::pass("No Unexpected / Missing Feeds in $desc") unless keys %has_feeds;

  my $failed = 0;
  for my $feed ( sort keys %has_feeds ) {
    if ( $has_feeds{$feed} > 0 ) {
      $failed++;
      Test::More::fail("$desc has unexpected extra $feed");
      next;
    }
    if ( $has_feeds{$feed} < 0 ) {
      $failed++;
      Test::More::fail("$desc is missing $feed");
      next;
    }
    Test::More::pass("$desc has feed $feed");
  }
  if ( $failed ) {
    Test::More::diag("\n---\nExpected feeds for $desc:");
    Test::More::diag(sprintf "- [%s]", join q[, ], sort @feeds );
    Test::More::diag("Got feeds for $desc:");
    Test::More::diag(sprintf "- [%s]", join q[, ], sort map { $_->href } $page->links('feed'));
    Test::More::diag("---");
  }
}

# -- no user servicable parts --
sub _class_functions {
  my ($class) = @_;
  Package::Stash->new($class)->list_all_symbols('CODE');
}

sub _function_type {
  my ($function) = @_;
  return 'PRIVATE'   if $function =~ /^_/;
  return 'TYPE_UTIL' if $function =~ /^(is_|assert_|to_)[A-Z]/;
  return 'PRIVATE'   if $function =~ /^[A-Z][A-Z]/;
  return 'TYPE'      if $function =~ /^[A-Z]/;
  return 'PUBLIC';
}

sub _hl_TYPE_UTIL {
  $_[0] =~ /^([^_]+_)(.*$)/;
  colored( \@TYPE_METHOD, $1 ) . colored( \@TYPE, $2 );
}
sub _hl_TYPE { colored( \@TYPE, $_[0] ) }
sub _hl_PUBLIC  { $_[1] ? colored( \@SHADOWED_PUBLIC,  $_[0] ) : colored( \@PUBLIC,  $_[0] ) }
sub _hl_PRIVATE { $_[1] ? colored( \@SHADOWED_PRIVATE, $_[0] ) : colored( \@PRIVATE, $_[0] ) }

sub _pp_function {
  return __PACKAGE__->can( '_hl_' . _function_type( $_[0] ) )->(@_);
}

sub _pp_key {
  my @tokens;
  push @tokens, "Public Function: " . _hl_PUBLIC("foo_example");
  push @tokens, "Type Constraint: " . _hl_TYPE("TypeName");
  push @tokens, "Type Constraint Utility: " . _hl_TYPE_UTIL("typeop_TypeName");
  push @tokens, "Private/Boring Function: " . _hl_PRIVATE("foo_example");
  if ($SHOW_SHADOWED) {
    push @tokens, "Public Function shadowed by higher scope: " . _hl_PUBLIC( "shadowed_example", 1 );
    push @tokens, "Private/Boring Function shadowed by higher scope: " . _hl_PRIVATE( "shadowed_example", 1 );
  }
  push @tokens, "No Functions: ()";
  return sprintf "Key:\n$INDENT%s\n\n", join qq[\n$INDENT], @tokens;
}

sub _mg_sorted {
  my (%functions) = @_;
  if ($SHOW_SHADOWED) {
    return ( [ sort { lc($a) cmp lc($b) } keys %functions ] );
  }
  return ( [ grep { !$functions{$_} } sort { lc($a) cmp lc($b) } keys %functions ] );
}

sub _mg_type_shadow_clustered {
  my (%functions) = @_;
  my %clusters;
  for my $function ( keys %functions ) {
    my $shadow = '.shadowed' x !!$functions{$function};
    $clusters{ _function_type($function) . $shadow }{$function} = $functions{$function};
  }
  my @out;
  for my $type ( map { $_, "$_.shadowed" } qw( PUBLIC PRIVATE TYPE TYPE_UTIL ) ) {
    next unless exists $clusters{$type};
    push @out, _mg_sorted( %{ $clusters{$type} } );
  }
  return @out;
}

sub _mg_type_clustered {
  my (%functions) = @_;
  my %clusters;
  for my $function ( keys %functions ) {
    $clusters{ _function_type($function) }{$function} = $functions{$function};
  }
  my @out;
  for my $type (qw( PUBLIC PRIVATE TYPE TYPE_UTIL )) {
    next unless exists $clusters{$type};
    push @out, _mg_sorted( %{ $clusters{$type} } );
  }
  return @out;
}

sub _mg_aleph {
  my (%functions) = @_;
  my %clusters;
  for my $function ( keys %functions ) {
    $clusters{ lc( substr $function, 0, 1 ) }{$function} = $functions{$function};
  }
  my @out;
  for my $key ( sort keys %clusters ) {
    push @out, _mg_sorted( %{ $clusters{$key} } );
  }
  return @out;

}

sub _pp_functions {
  my (%functions) = @_;
  my (@clusters)  = __PACKAGE__->can( '_mg_' . $CLUSTERING )->(%functions);
  my (@out_clusters);
  for my $cluster (@clusters) {
    my $cluster_out = '';

    my @functions = @{$cluster};
    while (@functions) {
      my $line = $INDENT;
      while ( @functions and length $line < $MAX_WIDTH ) {
        my $function = shift @functions;
        $line .= $function . q[, ];
      }
      $cluster_out .= "$line\n";
    }

    # Suck up trailing ,
    $cluster_out =~ s/,[ ]\n\z/\n/;
    $cluster_out =~ s{(\w+)}{ _pp_function($1, $functions{$1}) }ge;
    push @out_clusters, $cluster_out;
  }
  return join qq[\n], @out_clusters;
}

sub _pp_class {
  my ($class)        = @_;
  my $out            = q[];
  my $pad            = q[ ] x 4;
  my $seen_functions = {};
  for my $isa ( @{ mro::get_linear_isa($class) } ) {
    $out .= colored( ['green'], $isa ) . ":";
    my $section        = "";
    my $line           = $pad;
    my (@my_functions) = _class_functions($isa);
    if ( not @my_functions ) {
      $out .= " ()\n";
      next;
    }
    else { $out .= "\n" }
    my %function_map;
    for my $function (@my_functions) {
      if ( not exists $seen_functions->{$function} ) {
        $seen_functions->{$function} = $isa;
      }
      $function_map{$function} = ( $seen_functions->{$function} ne $isa );
    }
    $out .= _pp_functions(%function_map) . "\n";

    next;
  }
  return $out;
}

1;


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw( tunable ));
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

#$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
#$bundle->remove_policy('CodeLayout::RequireUseUTF8');
#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



lib/Statocles/AppRole/ExtraFeeds.pm
use 5.006;    # our
use strict;
use warnings;

package Statocles::AppRole::ExtraFeeds;

our $VERSION = '0.001004';

# ABSTRACT: Generate additional feed sets for apps

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo::Role qw( has around );
use Carp qw( croak );
use Statocles::App 0.070 ();    # 0.70 required for ->template
use Statocles::Page::List ();
use namespace::autoclean;

has 'extra_feeds' => (
  is      => 'ro',
  lazy    => 1,
  default => sub { {} },
);





use constant PATH_INDEX_PREFIX   => qr{  \A (.*) / index    [.](\w+) \z}sx;
use constant PATH_GENERIC_PREFIX => qr{  \A (.*) / ([^/.]+) [.](\w+) \z}sx;

# This separation is to help Coverage tools not lie to me that I did a perfect job
around pages => \&_around_pages;

sub _around_pages {
  my ( $orig, $self, @rest ) = @_;
  my (@pages) = $self->$orig(@rest);

  return @pages unless @pages;

  my %feed_cache;

  my @out_pages;

  while (@pages) {
    my $page = shift @pages;

    push @out_pages, $page;

    my (@existing_feeds) = $page->links('feed');

    next if not @existing_feeds;

    for my $feed_id ( sort keys %{ $self->extra_feeds } ) {
      my $feed = $self->extra_feeds->{$feed_id};
      my $feed_path;

      {
        my $reference_path = $existing_feeds[0]->href;
        my $feed_suffix = $feed->{name} || $feed_id;

        if ( $reference_path =~ PATH_INDEX_PREFIX ) {
          $feed_path = "$1/$feed_suffix";
        }
        elsif ( $reference_path =~ PATH_GENERIC_PREFIX ) {
          $feed_path = "$1/$2.$feed_suffix";
        }
        else {
          croak "Don't know how to derive feed path from $reference_path for $feed_suffix";
        }
      }

      if ( not exists $feed_cache{$feed_path} ) {
        my $feed_page = $feed_cache{$feed_path}{feed_page} = Statocles::Page::List->new(
          app      => $self,
          pages    => $page->pages,
          path     => $feed_path,
          template => $self->template( $feed->{template} || $feed->{name} || $feed_id ),
          links    => {
            alternate => [
              $self->link(
                href  => $page->path,
                title => ( $feed->{'index_title'} || $page->title || 'Web Index' ),
                type  => $page->type,
              ),
            ],
          },
        );

        $feed_cache{$feed_path}{feed_link} = $self->link(
          text => $feed->{text},
          href => $feed_page->path->stringify,
          type => $feed_page->type,
        );
      }
      $page->links( feed => $feed_cache{$feed_path}{feed_link} );
    }
  }
  return @out_pages, map { $feed_cache{$_}{feed_page} } sort keys %feed_cache;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Statocles::AppRole::ExtraFeeds - Generate additional feed sets for apps

=head1 VERSION

version 0.001004

=head1 EXPERIMENTAL

This module is very new and it comes with the following present caveats:

=over 4

=item * Application outside C<Statocles::App::Blog> untested.

Feedback welcome though, and it might work by magic!

=item * Implementation details a bit sketchy

The code works. I kinda feel like it shouldn't, and its like I've performed some magic
trick and the gods have smiled on me for a moment.

=item * You're on your own with templates

This at present is a glorified lump of glue on top of existing C<Statocles> behavior.

As such, if you want this to work, you'll probably want to copy some templates and modify them.

This module does nothing for you in terms of the actual formatting, it just pumps the right
glue so that the same code that generates the existing feeds will be invoked a few more times
but with the file names and templates you chose ( instead of the ones provided by default by the app )

Basically, you're going to want to copy C<blog/index.rss.ep> to C<blog/fulltext.rss.ep> and tweak
it a bit, or something.

=back

=head1 DESCRIPTION

This module is a role that can be applied to any C<Statocles::App> in a C<Statocles>'s C<Beam::Wire>
configuration.

  ...
  blog_app:
    class: 'Statocles::App::Blog'
    with: 'Statocles::AppRole::ExtraFeeds'
    args:
      url_root: '/blog'
      # ... more Statocles::App::Blog args
      extra_feeds:
        fulltext.rss:
          text: "RSS FullText"

This example creates a feed called C</blog/fulltext.rss> containing the contents of C<theme/blog/fulltext.rss.ep>
after template application, and is linked from every C<index> listing.

It also creates a feed called C<< /blog/tag/<%= tagname %>.fulltext.rss >> for each tag, provisioned from the same template.

=for Pod::Coverage PATH_INDEX_PREFIX PATH_GENERIC_PREFIX

=head1 PARAMETERS

=head2 C<extra_feeds>

This Role provides one tunable parameter on its applied class, C<extra_feeds>, which contains a
mapping of

  id => spec

=head3 C<extra_feeds> spec.

  {
    text      => required
    name      => default( id )
    template  => default( id )
  }

=head4 C<text>

This is the name of the feed when shown in links on both C<index> and C<tag index> pages.

=head4 C<template>

This is the name of the template to render the feeds content into.

Defaults to taking the same value as the key in the C<extra_fields> hash.

=head4 C<name>

This is the name of the file/file suffix that is generated.

It defaults to the same value as the key in the C<extra_feeds>
hash.

So:

  extra_feeds:
    fulltext.rss:
      text: "My FullText RSS"

And

  extra_feeds:
    genericlabel:
      text: "My FullText RSS"
      name: 'fulltext.rss'      # output name
      template: 'fulltext.rss'  # source template

Should both generate the same result.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Generate additional feed sets for apps",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "CPAN::Meta::Requirements" => "2.120620",
    "ExtUtils::MakeMaker" => 0,
    "Module::Metadata" => 0
  },
  "DISTNAME" => "Statocles-AppRole-ExtraFeeds",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Statocles::AppRole::ExtraFeeds",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Moo::Role" => 0,
    "Statocles::App" => "0.070",
    "Statocles::Page::List" => 0,
    "constant" => "1.03",
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Beam::Wire" => "1.016",
    "Data::Dump" => 0,
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Package::Stash" => 0,
    "Path::Tiny" => 0,
    "Scalar::Util" => 0,
    "Statocles::App::Blog" => 0,
    "Statocles::Test" => "0.075",
    "Term::ANSIColor" => 0,
    "Test::Deep" => 0,
    "Test::More" => 0,
    "constant" => "1.03",
    "lib" => 0
  },
  "VERSION" => "0.001004",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Beam::Wire" => "1.016",
  "Carp" => 0,
  "Data::Dump" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moo::Role" => 0,
  "Package::Stash" => 0,
  "Path::Tiny" => 0,
  "Scalar::Util" => 0,
  "Statocles::App" => "0.070",
  "Statocles::App::Blog" => 0,
  "Statocles::Page::List" => 0,
  "Statocles::Test" => "0.075",
  "Term::ANSIColor" => 0,
  "Test::Deep" => 0,
  "Test::More" => 0,
  "constant" => "1.03",
  "lib" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);

# inserted by Dist::Zilla::Plugin::DynamicPrereqs 0.032
# Interop for "with"
runtime_requires("Beam::Wire",1.016) if has_module("Beam::Wire","<1.016");


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

# inserted by Dist::Zilla::Plugin::DynamicPrereqs 0.032
sub _add_prereq {
  my ($mm_key, $module, $version_or_range) = @_;
  $version_or_range ||= 0;
  warn "$module already exists in $mm_key -- need to do a sane metamerge!"
    if exists $WriteMakefileArgs{$mm_key}{$module}
      and $WriteMakefileArgs{$mm_key}{$module} ne '0'
      and $WriteMakefileArgs{$mm_key}{$module} ne $version_or_range;
  warn "$module already exists in FallbackPrereqs -- need to do a sane metamerge!"
    if exists $FallbackPrereqs{$module} and $FallbackPrereqs{$module} ne '0'
        and $FallbackPrereqs{$module} ne $version_or_range;
  $WriteMakefileArgs{$mm_key}{$module} = $FallbackPrereqs{$module} = $version_or_range;
  return;
}

sub has_module {
  my ($module, $version_or_range) = @_;
  require Module::Metadata;
  my $mmd = Module::Metadata->new_from_module($module);
  return undef if not $mmd;
  return $mmd->version($module) if not defined $version_or_range;

  require CPAN::Meta::Requirements;
  my $req = CPAN::Meta::Requirements->new;
  $req->add_string_requirement($module => $version_or_range);
  return 1 if $req->accepts_module($module => $mmd->version($module));
  return 0;
}

sub runtime_requires {
  my ($module, $version_or_range) = @_;
  _add_prereq(PREREQ_PM => $module, $version_or_range);
}
